### PR TITLE
[Do not merge] Spike adr-8

### PR DESCRIPTION
**Pull request by alpe**
Design spike for discussion only.

Ref: 
* https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-008-app-caller-cbs/adr-008-app-caller-cbs.md
* https://github.com/cosmos/ibc-go/pull/3287

### CODE REVIEW

File: modules/apps/08-callbacks/ics20_example_cb.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	"encoding/json"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/ibc-go/v7/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
+)
+
+// MyICS20Payload my random payload example
+type MyICS20Payload struct {
```
Comment (alpe) [inline modules/apps/08-callbacks/ics20_example_cb.go:11]: Some example data that would be extracted from memo

File: modules/apps/08-callbacks/ics20_example_cb.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	// unmarshal json to type, I prefer using concrete type here:
+	var r MyICS20Payload
+	if err := json.Unmarshal([]byte(data.Memo), &r); err != nil {
+		return nil, err
+	}
+	// todo: check that contains valid data
+	return &r, nil
+}
+
+var _ Executor[MyICS20Payload] = &MyICS20Callbacks{}
```
Comment (alpe) [inline modules/apps/08-callbacks/ics20_example_cb.go:40]: Executor for ics20 only. my assumption was that chains may want to customize the callbacks for each of the ibc apps.

File: testing/simapp/app.go [hunk 3]
```diff
@@ -471,7 +472,12 @@ func NewSimApp(
 	// create IBC module from bottom to top of stack
 	var transferStack porttypes.IBCModule
 	transferStack = transfer.NewIBCModule(app.TransferKeeper)
-	transferStack = ibcfee.NewIBCMiddleware(transferStack, app.IBCFeeKeeper)
+	transferStack = callbacks.NewIBCMiddleware[callbacks.MyICS20Payload](
+		ibcfee.NewIBCMiddleware(transferStack, app.IBCFeeKeeper),
```
Comment (alpe) [inline testing/simapp/app.go:476]: This is how the middleware would be integrated

File: modules/apps/08-callbacks/ibc_middleware.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+	OnTimeoutPacketLimit         *sdk.Gas
+}
+
+// NewIBCMiddleware constructor
+func NewIBCMiddleware[T CallbackPacketData](next porttypes.Middleware, decoder Decoder[T], executor Executor[T], limits GasLimits) IBCMiddleware[T] {
+	return IBCMiddleware[T]{
+		next:     next,
+		decoder:  decoder,
+		executor: executor,
+		limits:   limits,
```
[the remark this answers is not in the document]
Comment (alpe) [reply modules/apps/08-callbacks/ibc_middleware.go:52]: Adr-8 defines the limits read from the packet not defined here. So this is wrong

**Comment by colin-axner** [comment]
Thanks @alpe :pray: I'm going to close this pr since there won't be any more work done for it and the design discussion has progressed past it.

**Pull request closed without merging**